### PR TITLE
feat: depth estimation tutorial.

### LINKED ISSUE #263: Add fine-tuning tutorial for depth estimation
Cc: @NielsRogge @osanseviero @nateraw

### PR DESCRIPTION
**Pull request by sayakpaul**
Closes #263

### CODE REVIEW

**Comment by sayakpaul** [comment]
Sorry for the delay on this one. I am trying hard to get some reasonable results here. Have run a couple of experiments, but the results are wonky (at least to me). 

To this end, I am maintaining my experimentation code here: https://github.com/sayakpaul/depth_estimation_trainer. 

The corresponding WandB project is here: https://wandb.ai/sayakpaul/depth_estimation. Note that it does not reflect all the experiments I have conducted so far. Will run a couple more tomorrow. 

Meanwhile, I would really appreciate 👀 into the codebase (a [single file](https://github.com/sayakpaul/depth_estimation_trainer/blob/main/run_depth_estimation.py) really) of the above-mentioned repo. 

@NielsRogge @nateraw

**Comment by sayakpaul** [comment]
<p>The original training subset is 82 GBs in size. That is why I wanted to have a subset of it (1715 images subsampled). The original validation set is &lt; 3 GB having ~300 images. So, wanted to keep it as is. </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/264/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
@NielsRogge @nateraw a gentle ping ^

**Comment by sayakpaul** [comment]
@NielsRogge @nateraw a gentle reminder on this: https://github.com/huggingface/notebooks/pull/264#issuecomment-1316916104. 

So far, I have tried the following things:

* Starting from the pre-trained decoder and depth head alongside the encoder, so far performs the best.  
* Randomly initializing the decoder and depth head doesn't lead to greater performance. 
* Using augmentation transforms doesn't help with the performance. 
* Training for longer or changing the hyperparameter do not seem to have that much of an impact on the performance. 

W B Chart 14_12_2022, 20_42_53
<small><a href="https://wandb.ai/sayakpaul/depth_estimation">WandB Dashboard</a></small>

Visualized results are as follows (also available from the dashboard under the "Media" tab):

**Model initialized with `vinvino02/glpn-kitti`**:

image

**Model initialized with `vinvino02/glpn-nyu`**:

image

Both the above training runs share the same training hyperparameters except the starting checkpoint. 

From the visuals, it looks like initializing from `vinvino02/glpn-nyu` helps a little but the results seem quite off to my eyes. 

Would appreciate your inputs.

**Comment by sayakpaul** [comment]
> @sayakpaul The code in your repo looks good to me, besides your issue with results. Did you want us to actually try and run to get better results, or just review the code (I did the latter). Lmk what I can do to help 😄

@nateraw thanks! Primarily, I wanted your feedback on the code -- if I was doing something noticeably wrong. @NielsRogge suggested trying to overfit a single batch which I will do today and keep this thread updated.

**Comment by sayakpaul** [comment]
@nateraw some updates on the single batch overfitting experiments:

The setup: 6 training samples and the same samples as the validation examples. Batch size = 6. Number of epochs = 100. Learning rate: 2e-5 without any warmup ratio, keeping the rest of the arguments in `TrainingArguments` to their default values. 

* Initializing from `vinvino02/glpn-nyu`:

pretrained

* Randomly initialized network:

random_one

But to amortize the random initialization, when I changed the learning rate to 2e-4 and warmup ratio to 0.2 (keeping rest of the things the same), I get:

random_two

Cannot say the results are amazing but changing the hyperparameters here seemed to have at least improved the previous results (when the network is randomly initialized). 

But from the ground-truth labels, there seem to be issues with the depth maps i.e., incorrect measurements of depths. This might be influencing the results. 

Any experiments you want to suggest? Or any other general comments?

**Comment by NielsRogge** [comment]
I'd suggest taking a look at the GLPN authors, see which hyperparameters they used.

I see that after 100 epochs the loss is still far above 0, so it might have to do with learning rate. Normally loss should be going down quickly to 0

**Comment by sayakpaul** [comment]
> I'd suggest taking a look at the GLPN paper, see which hyperparameters they used.

GLPN paper doesn't include any experiments on transfer learning. Are you suggesting to follow their pre-training hparams?

> I see that after 100 epochs the loss is still far above 0, so it might have to do with learning rate. Normally loss should be going down quickly to 0

Good one. Will try with a couple more LRs.

**Comment by sayakpaul** [comment]
@radames the PR is still not in a ready state, hence.

**Comment by radames** [comment]
I'm so sorry @sayakpaul I think I accidentally did that 😳

**Comment by sayakpaul** [comment]
Closing it because we already have a guide on depth estimation. https://huggingface.co/docs/transformers/main/tasks/monocular_depth_estimation
